processing/src/main/java/org/apache/druid/query/dimension/PrefixFilteredDimensionSpec.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.query.dimension;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Predicate;
+import org.apache.druid.common.config.NullHandling;
+import org.apache.druid.java.util.common.StringUtils;
+import org.apache.druid.query.filter.DimFilterUtils;
+import org.apache.druid.segment.DimensionSelector;
+import it.unimi.dsi.fastutil.ints.Int2IntMap;
+import it.unimi.dsi.fastutil.ints.Int2IntOpenHashMap;
+import java.nio.ByteBuffer;
+import javax.annotation.Nullable;
+
+/**
+ */
+public class PrefixFilteredDimensionSpec extends BaseFilteredDimensionSpec
+{
+
+  private static final byte CACHE_TYPE_ID = 0x4;
+
+  private final String prefix;
+
+  public PrefixFilteredDimensionSpec(
+      @JsonProperty("delegate") DimensionSpec delegate,
+      @JsonProperty("prefix") String prefix //rows not starting with the prefix will be discarded
+  )
+  {
+    super(delegate);
+    this.prefix = Preconditions.checkNotNull(prefix, "prefix must not be null");
+  }
+
+  @JsonProperty
+  public String getPrefix()
+  {
+    return prefix;
+  }
+
+  @Override
+  public DimensionSelector decorate(final DimensionSelector selector)
+  {
+    if (selector == null) {
+      return null;
+    }
+
+    final int selectorCardinality = selector.getValueCardinality();
+    if (selectorCardinality < 0 || !selector.nameLookupPossibleInAdvance()) {
+      return new PredicateFilteredDimensionSelector(
+          selector,
+          new Predicate<String>()
+          {
+            @Override
+            public boolean apply(@Nullable String input)
+            {
+              return NullHandling.nullToEmptyIfNeeded(input).startsWith(prefix);
```

Comment:
The result of `NullHandling.nullToEmptyIfNeeded(input)` needs a null check, since it can return null if `druid.generic.useDefaultValueForNull` is false

Reply:
The same issue with `RegexFilteredDimensionSpec`, can be modified along with this `PrefixFilteredDimensionSpec`.